### PR TITLE
Add microseconds to the TimeUnit enum

### PR DESCRIPTION
**Pull request by tomerd**
Motivation:

I have a metrics backend where I need to store timer values in
microseconds. The "preferred display unit" models this exactly, but it
seems to have (inadvertently?) omitted the possibility to express
microseconds. Note that among the timer "report" methods there is
already a microseconds option.

Modifications:

adding an emum case is an API breaking change. so instead of just adding the enum case for microseconds, we use this opportunity to change `TimeUnit` from enum to struct so that we can add more cases in the future without breaking the API.

### CODE REVIEW

**Comment by tomerd** [comment]
to replace https://github.com/apple/swift-metrics/pull/62

File: Tests/MetricsTests/TestMetrics.swift [hunk 1]
```diff
@@ -166,7 +166,9 @@ internal class TestTimer: TimerHandler, Equatable {
             case .minutes: return (value / 1_000_000_000) * 60
             case .seconds: return value / 1_000_000_000
             case .milliseconds: return value / 1_000_000
+            case .microseconds: return value / 1000
             case .nanoseconds: return value
+            default: preconditionFailure("unknown display unit: \(displayUnit)")
```
Comment (weissi) [inline Tests/MetricsTests/TestMetrics.swift]: Should we maybe store a conversion to another unit for every new unit? We could for example require that each new unit defines its conversion to nanoseconds so we don’t have to crash if we don’t know it?

Comment (MrLotU) [reply Tests/MetricsTests/TestMetrics.swift]: Or perhaps as a `default` return `value` without any conversion?

Comment (weissi) [reply Tests/MetricsTests/TestMetrics.swift]: @MrLotU that would lead to wrong data, no?

Comment (MrLotU) [reply Tests/MetricsTests/TestMetrics.swift]: Ah. Of course 🤦‍♂ I think in that case your proposed solution is a lot better. Disregard my previous comment.

Comment (cburrows) [reply Tests/MetricsTests/TestMetrics.swift]: I agree with @weissi , this struct might as well be a Double (or Int64) that expresses how many are in a nanosecond. And if you expose that value then no one has to care about the addition of well named static units ever again.

Comment (weissi) [reply Tests/MetricsTests/TestMetrics.swift]: Double is a good idea. We could also make “seconds” (as the SI unit) the base and require everything to be able to express itself in seconds if nothing better is available?

Comment (tomerd) [reply Tests/MetricsTests/TestMetrics.swift]: good idea: https://github.com/apple/swift-metrics/pull/64/commits/1c4103e92d52b05e356e66e3b1dd2b51b3caba0b

File: Sources/CoreMetrics/Metrics.swift [hunk 2]
```diff
@@ -183,19 +183,23 @@ public struct TimeUnit: Equatable {
         case days
     }
 
-    private var code: Code
+    private let code: Code
+    let scaleFromNanoseconds: UInt
```
Comment (weissi) [inline Sources/CoreMetrics/Metrics.swift]: This needs to be `UInt64` for 32 bit platforms (or Double)

File: Sources/CoreMetrics/Metrics.swift [hunk 3]
```diff
@@ -172,8 +172,34 @@ public class Gauge: Recorder {
...
+        case microseconds
+        case milliseconds
+        case seconds
+        case minutes
+        case hours
+        case days
+    }
+
+    private let code: Code
+    let scaleFromNanoseconds: UInt64
```
Comment (cburrows) [inline Sources/CoreMetrics/Metrics.swift]: Should this be `public`? Otherwise no backing metrics module can use it and would need to know the names.

Comment (cburrows) [reply Sources/CoreMetrics/Metrics.swift]: Also good call, `Double` would be weird with `Equatable`.

File: Sources/CoreMetrics/Metrics.swift [hunk 4]
```diff
@@ -172,8 +172,34 @@ public class Gauge: Recorder {
...
+        case nanoseconds
+        case microseconds
+        case milliseconds
+        case seconds
+        case minutes
+        case hours
+        case days
+    }
+
+    private let code: Code
```
Comment (ktoso) [inline Sources/CoreMetrics/Metrics.swift]: code reads a bit weird, `encoding`?
```suggestion
    private let encoding: Encoding
```

?

**Comment by tomerd** [comment]
@cburrows @tanner0101 @tachyonics @MrLotU @Mordil @Blejwi @ktoso tagged as 2.0.0

**Pull request merged**